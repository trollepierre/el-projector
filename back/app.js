var createError = require('http-errors')
var express = require('express')
var bodyParser = require('body-parser')
var path = require('path')
var cookieParser = require('cookie-parser')
var logger = require('morgan')
const cors = require('cors')

var tasksRouter = require('./src/routes/tasks-routes');
var dbRouter = require('./src/routes/db');

var app = express()

app.use(logger('dev'))
app.use(bodyParser.json())
app.use(bodyParser.urlencoded({ extended: false }))
app.use(cookieParser())
app.use(cors())

app.use(express.static(path.join(__dirname, '..', 'front-react', 'build')))

app.use('/tasks', tasksRouter);
app.use('/db', dbRouter);

// catch 404 and forward to error handler
app.use(function (req, res, next) {
  next(createError(404))
})

// error handler
app.use(function (err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message
  res.locals.error = req.app.get('env') === 'development' ? err : {}

  // render the error page
  res.status(err.status || 500)
})

module.exports = app
