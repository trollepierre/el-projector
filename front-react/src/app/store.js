import { withReducer } from '../store/withReducer';
import { apiService, tokenService } from '../services';

const SET_IS_AUTHENTICATED = 'app/SET_IS_AUTHENTICATED';

const setIsAuthenticated = isAuthenticated => dispatch => {
  if(!isAuthenticated){
    tokenService.removeTokens()
  }
  return dispatch({
    type: SET_IS_AUTHENTICATED, payload: { isAuthenticated }
  });
};

const authenticate = password => async dispatch => {
  try {
    const data = {
      email: 'some@example.org',
      name: password
    };
    const tokens = await apiService.post('login', data)
    await tokenService.saveUserTokens(tokens, data)
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    alert('wrong password')
  }
}

const loginSilently = () => async dispatch => {
  try {
    const refreshToken = tokenService.getRefreshToken();
    if(!refreshToken) return;
    const tokens = await apiService.post('login/token', { refreshToken })
    await tokenService.reauthenticate(tokens);
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    await tokenService.removeTokens()
    alert('problem during login silently')
    throw error
  }
}

export const actions = {
  setIsAuthenticated,
  authenticate,
  loginSilently,
};

export const initialState = {
  isAuthenticated: false,
};

export const actionHandlers = {
  [SET_IS_AUTHENTICATED]: (state, action) => {
    const { isAuthenticated } = action.payload;
    return {
      ...state,
      isAuthenticated,
    };
  },
};

export const reducer = withReducer(initialState, actionHandlers);
