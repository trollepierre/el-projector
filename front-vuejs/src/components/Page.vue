<template>
  <div class="page">
    <TaskTable />
  </div>
</template>

<script>
import TaskTable from './TaskTable.vue'

export default {
  name: 'HelloWorld',
  components: { TaskTable },
}
</script>

<style scoped>
h1 {
  font-weight: normal;
}
</style>
