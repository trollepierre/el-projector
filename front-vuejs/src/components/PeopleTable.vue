<template>
  <div class="my-table">
    <table>
      <TableHeader
        @sortBirthday="sortBirthday"
        @sortByFirstName="sortByFirstName"
        @sortByLastDate="sortByLastDate"
        @sortByLastName="sortByLastName"
        @sortByNext="sortByNext"
        @sortByOrigin="sortByOrigin"
        @sortByPoints="sortByPoints"
      />
      <tbody>
        <CreateTask
          :new-task="newTask"
          @createNewTask="create"
        />
        <div
          v-for="task in sortedTask"
          :key="task.id"
        >
          <template v-if="isEditable(task)">
            <EditTask
              :task="task"
              @deleteItem="deleteItem"
              @save="save"
            />
          </template>
          <template v-else>
            <Task
              :birthday="isBirthday(task)"
              :task="task"
              @meet="meet"
              @differ="differ"
              @reject="reject"
              @editItem="editItem"
              @deleteItem="deleteItem"
            />
          </template>
        </div>
      </tbody>
    </table>
  </div>
</template>

<script>
  import { equals } from 'ramda'
  import tasksApi from '../api/tasks'
  import meetingApi from '../api/meeting'
  import tasksSorter from '../services/tasks-sorter'
  import TableHeader from './table-header/TableHeader.vue'
  import CreateTask from './create-task/CreateTask.vue'
  import EditTask from './edit-task/EditTask.vue'
  import Task from './task/Task.vue'

  export default {
    name: 'TaskTable',
    components: { Task, EditTask, CreateTask, TableHeader },
    data() {
      return {
        tasks: [],
        editTask: {},
        newTask: {},
      }
    },
    computed: {
      sortedTask() {
        return this.tasks
      },
    },
    mounted() {
      this.fetchTask()
    },
    methods: {
      isEditable(task) {
        return equals(task, this.editTask)
      },
      async fetchTask() {
        this.tasks = await tasksApi.fetch()
      },
      async meet(metTask) {
        const updatedInformations = await meetingApi.meet(metTask)
        this.updateInformation(updatedInformations, metTask.id)
      },
      async differ(task) {
        const updatedInformations = await meetingApi.differ(task)
        this.updateInformation(updatedInformations, task.id)
      },
      async reject(task) {
        const updatedInformations = await meetingApi.reject(task)
        this.updateInformation(updatedInformations, task.id)
      },
      async deleteItem(task) {
        if (confirm('Are you sure you want to delete this item?')) {
          await tasksApi.delete(task)
          this.tasks = this.tasks.filter(({ id }) => id !== task.id)
        }
      },
      editItem(task) {
        this.editTask = task
      },
      async save(task) {
        const updatedInformations = await tasksApi.update(task)
        this.updateInformation(updatedInformations, task.id)
        this.sortByNext()
        this.reset()
      },
      reset() {
        this.newTask = {}
        this.editTask = {}
      },
      async create(task) {
        const createdTask = await tasksApi.create(task)
        this.tasks.push(createdTask)
        this.reset()
        this.sortByNext()
      },
      // TODO refacto this - updateInformation (updatedInformations, idTask, tasks) {
      updateInformation(updatedInformations, id) {
        this.tasks = this.tasks.map(task => {
          if (task.id === id) {
            return { ...task, ...updatedInformations }
          }
          return task
        })
      },
      sortByNext() {
        this.tasks = tasksSorter.next(this.tasks)
      },
      sortByFirstName() {
        this.tasks = tasksSorter.firstName(this.tasks)
      },
      sortByLastName() {
        this.tasks = tasksSorter.lastName(this.tasks)
      },
      sortByPoints() {
        this.tasks = tasksSorter.points(this.tasks)
      },
      sortByLastDate() {
        this.tasks = tasksSorter.lastDate(this.tasks)
      },
      sortByOrigin() {
        this.tasks = tasksSorter.origin(this.tasks)
      },
      sortBirthday() {
        this.tasks = tasksSorter.birthday(this.tasks)
      },
      isBirthday(task) {
        return task.next === task.nextBirthday ? 'birthday' : ''
      },
    },
  }
</script>

<style scoped>
  .my-table {
    padding-top: 30px;
  }

  table {
    width: 100%;
  }
</style>
