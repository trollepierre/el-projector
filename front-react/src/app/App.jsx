import React from 'react';
import { Provider } from 'react-redux';
import ProjectPage from '../pages/ProjectPage/ProjectPage';
import { default as createStore } from '../store/createStore';
import { default as reducer } from '../bricks/project-table/project-body/store';
import LoginPage from '../pages/LoginPage';
import { auth } from '../services';

const store = createStore(reducer)

const App = () => {
  if(!auth.isAuthenticated()) return (<LoginPage/>)

  return (
    <div className="App">
      <Provider store={store}>
        <ProjectPage/>
      </Provider>
    </div>
  );
};

export default App;
