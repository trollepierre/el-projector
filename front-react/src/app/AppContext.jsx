import React, { useReducer, createContext, useContext } from 'react';
import PropTypes from 'prop-types';
import { reducer, initialState, actions } from './store';

const AppContext = createContext();

export const useAppContext = () => useContext(AppContext);

const getProps = (state, dispatch) => ({
  ...state,
  setIsAuthenticated: isAuthenticated => actions.setIsAuthenticated(isAuthenticated)(dispatch),
  authenticate: value => actions.authenticate(value)(dispatch),
});

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <AppContext.Provider value={{ ...getProps(state, dispatch) }}>
      {children}
    </AppContext.Provider>
  );
};

AppProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default AppContext;
