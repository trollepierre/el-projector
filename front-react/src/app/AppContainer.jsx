import React, { useEffect } from 'react';
import { AppProvider, useAppContext } from './AppContext';

import LoginPage from '../pages/LoginPage';
import { Header } from '../bricks';
import ProjectPage from '../pages/ProjectPage/ProjectPage';

const AppContainer = () => {
  const { isAuthenticated } = useAppContext();

  if (!isAuthenticated) {
    return (<LoginPage/>);
  }

  return (
    <div className="App">
      <Header/>
      <ProjectPage/>
    </div>
  );
};

const AppModule = () => {
  return (
    <AppProvider>
      <AppContainer/>
    </AppProvider>
  );
};

export default AppModule;
