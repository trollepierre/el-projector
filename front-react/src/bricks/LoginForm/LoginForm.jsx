import React from 'react';
import { Button, Input } from '../../components';
import { useAppContext } from '../../app/AppContext';

const LoginForm = () => {
  const { authenticate } = useAppContext()
  const input = React.createRef();

  const handleSubmit = event => {
    event.preventDefault()
    authenticate(input.current.value)
  }

  return (
    <form className="login-form" onSubmit={handleSubmit}>
      <Input type='text' name={'mot de passe'} ref={input}/>
      <Button type='submit' onClick={handleSubmit} height={100} text='Se Connect'>Se connecter</Button>
    </form>
  );
};

export default LoginForm;
