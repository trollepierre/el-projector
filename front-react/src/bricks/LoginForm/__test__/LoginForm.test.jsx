import React from 'react';
import LoginForm from '../LoginForm';
import { shallow } from 'enzyme';
import * as AppContext from '../../../app/AppContext';
import { Button } from '../../../components';

describe('LoginForm', () => {
  let authenticateMock;

  beforeEach(() => {
    authenticateMock = jest.fn();
    jest
      .spyOn(AppContext, 'useAppContext')
      .mockImplementation(() => ({ authenticate: authenticateMock }));
  });

  it('should match snapshot', () => {
    // When
    const wrapper = shallow(<LoginForm/>);

    // Then
    expect(wrapper).toMatchSnapshot();
  });

  describe('events', () => {
    describe('onClick on logout button', () => {
      beforeEach(() => {
        jest
          .spyOn(React, 'createRef')
          .mockImplementation(() => ({ current: {value: 'value'} }));
      })

      it('should authenticate with value', () => {
        // Given
        const wrapper = shallow(<LoginForm/>);
        const event = { preventDefault: jest.fn()}

        // When
        wrapper.find(Button).simulate('click', event);

        // Then
        expect(authenticateMock).toHaveBeenCalledWith('value');
      });

      it('should prevent event default', () => {
        // Given
        const wrapper = shallow(<LoginForm/>);
        const event = { preventDefault: jest.fn()}

        // When
        wrapper.find(Button).simulate('click', event);

        // Then
        expect(event.preventDefault).toHaveBeenCalledWith();
      });
    });

    describe('onSubmit of form', () => {
      beforeEach(() => {
        jest
          .spyOn(React, 'createRef')
          .mockImplementation(() => ({ current: {value: 'value'} }));
      })

      it('should authenticate with value', () => {
        // Given
        const wrapper = shallow(<LoginForm/>);
        const event = { preventDefault: jest.fn()}

        // When
        wrapper.find('.login-form').simulate('submit', event);

        // Then
        expect(authenticateMock).toHaveBeenCalledWith('value');
      });

      it('should prevent event default', () => {
        // Given
        const wrapper = shallow(<LoginForm/>);
        const event = { preventDefault: jest.fn()}

        // When
        wrapper.find('.login-form').simulate('submit', event);

        // Then
        expect(event.preventDefault).toHaveBeenCalledWith();
      });
    });

  });
});
