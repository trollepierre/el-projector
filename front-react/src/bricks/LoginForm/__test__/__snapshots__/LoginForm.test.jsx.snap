// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match snapshot 1`] = `
<form
  className="login-form"
  onSubmit={[Function]}
>
  <Input
    dataQa=""
    disabled={false}
    error=""
    helper=""
    hidden={false}
    label=""
    name="mot de passe"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    required={false}
    reset={false}
    reverse={false}
    spellCheck={false}
    touched={true}
    type="text"
    valid={false}
    value=""
  />
  <Button
    height={100}
    onClick={[Function]}
    text="Se Connect"
    type="submit"
  >
    Se connecter
  </Button>
</form>
`;
